### PR TITLE
Valeriyapetrova/hometask 4

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Comparison methods rotate and lab_rotate

       user     system      total        real
"Comarison on 100"
my method  0.000062   0.000016   0.000078 (  0.000061)
real method  0.000015   0.000001   0.000016 (  0.000015)
my method  0.000034   0.000002   0.000036 (  0.000035)
real method  0.000012   0.000001   0.000013 (  0.000012)
my method  0.000021   0.000001   0.000022 (  0.000021)
real method  0.000023   0.000003   0.000026 (  0.000018)
"Comarison on 1000"
my method  0.001911   0.000039   0.001950 (  0.005556)
real method  0.000062   0.000002   0.000064 (  0.000057)
my method  0.000200   0.000019   0.000219 (  0.000198)
real method  0.000083   0.000010   0.000093 (  0.000094)
my method  0.000216   0.000003   0.000219 (  0.000352)
real method  0.000102   0.000015   0.000117 (  0.000131)
"Comarison on 10_000"
my method  0.001025   0.000228   0.001253 (  0.002521)
real method  0.000673   0.000121   0.000794 (  0.002074)
my method  0.001713   0.000176   0.001889 (  0.002249)
real method  0.000715   0.000150   0.000865 (  0.000865)
my method  0.001825   0.000183   0.002008 (  0.003530)
real method  0.000410   0.000077   0.000487 (  0.000479)

**Comment by ghost** [comment]
Comparison Hash and MyHash

user     system      total        real
"Hash"
   0.000016   0.000008   0.000024 (  0.000007)
   0.000003   0.000001   0.000004 (  0.000003)
   0.000009   0.000002   0.000011 (  0.000006)
"MyHash"
   0.000013   0.000001   0.000014 (  0.000012)
   0.000004   0.000001   0.000005 (  0.000004)
   0.000003   0.000000   0.000003 (  0.000003)
       user     system      total        real
"Hash determines size"
   0.000025   0.000001   0.000026 (  0.000025)
   0.000157   0.000013   0.000170 (  0.000179)
   0.001691   0.000181   0.001872 (  0.002020)
   0.005741   0.000137   0.005878 (  0.006879)
   0.000168   0.000026   0.000194 (  0.000309)
   0.027820   0.001074   0.028894 (  0.033270)
"Hash delete everything"
   0.000045   0.000008   0.000053 (  0.000041)
   0.000134   0.000001   0.000135 (  0.000133)
   0.001408   0.000001   0.001409 (  0.001412)
   0.000137   0.000002   0.000139 (  0.000137)
   0.001219   0.000142   0.001361 (  0.001632)
   0.020124   0.000995   0.021119 (  0.022053)

File: valeriyapetrova/hometask-4/my_hash.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @current_custom_hash[key_hash(key_input) % @size].delete(pair)
+  end
+
+  def size_of_my_hash
+    @count
+  end
+
+  def cleaning
+    @current_custom_hash.clear
+    @size = DEFAULT_MY_HASH_SIZE
```
Comment (TanyaAkulich) [inline valeriyapetrova/hometask-4/my_hash.rb]: the code is the same like in `initialize`. Let's follow DRY

File: valeriyapetrova/hometask-4/my_hash.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    end
+  end
+
+  def [](key_input)
+    pair = @current_custom_hash[key_hash(key_input) % @size]&.find { |b| b[0] == key_input }
+    pair ? pair[1] : nil
+  end
+
+  def delete(key_input)
+    pair = @current_custom_hash[key_hash(key_input) % @size]&.find { |b| b[0] == key_input }
```
Comment (TanyaAkulich) [inline valeriyapetrova/hometask-4/my_hash.rb]: code here and in `def [](key_input)` and `def []=(key_input, value_input)`  are very similar, let's move out the common

File: valeriyapetrova/hometask-4/my_hash.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @size = DEFAULT_MY_HASH_SIZE
+    @count = INIT_MY_HASH_COUNT
+    @current_custom_hash = Array.new(DEFAULT_MY_HASH_SIZE) { [] }
+  end
+
+  def []=(key_input, value_input)
+    resize if (@count.to_f / @size) > CAPACITY
+    pair = @current_custom_hash[key_hash(key_input) % @size]&.find { |b| b[0] == key_input }
+    if pair
+      pair[1] = value_input
```
Comment (TanyaAkulich) [inline valeriyapetrova/hometask-4/my_hash.rb]: 0 and 1 are a magic numbers here, maybe it's better move it to constants

File: valeriyapetrova/hometask-4/my_hash.rb (hunk 2 above)
Comment (TanyaAkulich) [reply valeriyapetrova/hometask-4/my_hash.rb]: let's make `@current_custom_hash[key_hash(key_input) % @size]` expression more readable

**Review by TanyaAkulich — CHANGES_REQUESTED** [review]
order the commit history

**Review by TanyaAkulich — APPROVED** [review]

**Pull request merged**